src/Microsoft.Security.Utilities.Core/PreciselyClassifiedSecurityKeys/SEC101_031_NuGetApiKey.cs

```diff
@@ -1,24 +1,37 @@
 ﻿// Copyright (c) Microsoft. All rights reserved.
 // Licensed under the MIT license. See LICENSE file in the project root for full license information.
-using System;
+
 using System.Collections.Generic;
 
 namespace Microsoft.Security.Utilities
 {
     public class NuGetApiKey : RegexPattern
     {
+        private const string Base32 = "abcdefghijklmnopqrstuvwxyz234567";
+
         public NuGetApiKey()
         {
             Id = "SEC101/031";
             Name = nameof(NuGetApiKey);
-            DetectionMetadata = DetectionMetadata.FixedSignature | DetectionMetadata.HighEntropy| DetectionMetadata.HighConfidence;
-            Pattern = "(^|[^0-9a-z])(?<refine>oy2[a-p][0-9a-z]{15}[aq][0-9a-z]{11}[eu][bdfhjlnprtvxz357][a-p][0-9a-z]{11}[aeimquy4])([^aeimquy4]|$)";
+            DetectionMetadata = DetectionMetadata.FixedSignature | DetectionMetadata.HighEntropy | DetectionMetadata.HighConfidence;
+
+            // This is the ApiKeyV4 format implemented here:
+            // https://github.com/NuGet/NuGetGallery/blob/main/src/NuGetGallery.Services/Authentication/ApiKeyV4.cs
+            Pattern = "(^|[^a-zA-Z0-9])(?<refine>oy2[a-z2-7]{43})([^a-zA-Z0-9]|$)";
```

Comment:
>(^|[^a-zA-Z0-9])(?<refine>oy2[a-z2-7]{43})([^a-zA-Z0-9]|$)" [](http://example.com/codeflow?start=23&length=59)

[https://github.com/search?q=%2F%28%3F-i%29%28%5E%7C%5B%5Ea-z0-9%2B%5C%2F%5C-_%5D%29oy2%5Ba-z2-7%5D%7B43%7D%28%24%7C%5B%5Ea-z0-9%2B%5C%2F%5C-_%5D%29%2F&type=code](https://github.com/search?q=%2F%28%3F-i%29%28%5E%7C%5B%5Ea-z0-9%2B%5C%2F%5C-_%5D%29oy2%5Ba-z2-7%5D%7B43%7D%28%24%7C%5B%5Ea-z0-9%2B%5C%2F%5C-_%5D%29%2F&type=code)

Playing around with this in GH, I think we should go ahead and add all base64 encoding relevant chars, i.e. Let me know what you think.

(?-i)(^|[^a-z0-9+\/\-_])oy2[a-z2-7]{43}([^a-z0-9+\/\-_]|$)

Reply:
This looks reasonable to me! I'll add it to the PR.

Reply:
This does not account for you concern that a user might have a leading or trailing hyphen separating the token from other data. Just want to acknowledge that. I just didn't feel like it made sense to exclude only this character. It 'feels' better to add all four special base64 characters. Alternately, we could simply include `+` and `/` (and not have special delimiter logic for the hyphen and underscore. I suspect this latter approach would be just fine and may in fact be the safest change. I leave it to you, your thoughts?

Reply:
In my opinion, if there is any lack of clarity at this stage, we should take the reasonable option that has the lowest chance of missing a true positive. Along those lines, a NuGet API key bounded by alphanumeric characters seems reasonable to miss because I can't think of any config file or script I've used in the past where a string bound is `[a-z0-9]`, whereas `+`, `/`, `-`, or `_` all feel like visual boundaries on a string different from alphanumeric (although definitely less common than say a whitespace character or a quote character).

I think your mention of case insensitivity is important since, upon inspection, it looks like our code is actually case _insensitive_ on API keys (!!!). Meaning `OY2...` is accepted by our endpoint. I just tested it. We do hand out lowercase base32 when a user creates an API key but if for some crazy reason if they uppercase their API key in their config file/script, it would still work. https://github.com/NuGet/NuGetGallery/blob/7c81548db51ed543ff8f6558c1306571a6bb368a/src/NuGetGallery.Services/Authentication/ApiKeyV4.cs#L124-L126

So if it is left to me, I would do this:
```
(?i)(^|[^a-z0-9])oy2[a-z2-7]{43}([^a-z0-9]|$)
```

![image](https://github.com/user-attachments/assets/e6f6b1fb-82a2-43d5-9da9-4ac4cd8651cc)
https://regex101.com/r/GALkvJ/1

Thus we are accepting that some unlucky base64 will be considered as a NuGet API key and will get flagged (and sent to our secret invalidation endpoint). If we see this cause problems in the future in any usage of this library or with GitHub Secret Scanning, we can reassess this decision with that clear case in mind.

Reply:
As a meta point, I believe this decision is a not a very scary one (between any of the regexes we have come up with) because we are wanting to move to HISv2 anyway :)

Reply:
Alright, sorry to rathole on this. This library is used for redaction purposes, and so it is actually a concern if we remove unlucky base64 data. We do see evidence of noisy hits in the wild, the question is whether we will comfortable redacting that data from, say, a build log or telemetry stream. 

For sure, GitHub should receive a looser version, in that context, no data is removed, any finding can be bypassed, and any reported finding can be easily filtered by the handler (you, NuGet).

So, this is getting a bit detailed, but one thing we could consider doing is this: we proceed with the pattern that you've suggested, which may find an unlucky base64 pattern embedded in other data. We have a minor bit of post-processing that examines our find and in the case that we determine the casing of the token is inconsistent (i.e., we see a mix of upper and lower case), we reject the find as unlucky base64 data.

Now, if your back-end is completely case-insensitive, a user could mix casing arbitrarily. But it seems like it could be a rare occurrence someone would partly modify a token in this way. 

It is non-obvious but `GetMatchIdAndName` allows for this post-processing. This helper receives a match. If you reject it, return null, otherwise let the base class handle things (it will simply construct moniker for the find, which is the combined rule id + friendly name).

See this code for an example:

[https://github.com/microsoft/security-utilities/blob/034040c8038c27dbac3bbe1fd8e2bcec37f0ffa2/src/Microsoft.Security.Utilities.Core/PreciselyClassifiedSecurityKeys/SEC101_102_AdoPat.cs#L22](https://github.com/microsoft/security-utilities/blob/034040c8038c27dbac3bbe1fd8e2bcec37f0ffa2/src/Microsoft.Security.Utilities.Core/PreciselyClassifiedSecurityKeys/SEC101_102_AdoPat.cs#L22)

Reply:
Ratholing on things like this is an indulgence of mine. Thank you for the context, this helps a lot! I see now that there are multiple distinct use cases of secret detection and the pattern accordingly may vary. I will look at the `GetMatchIdAndName` method and try to enhance it for NuGet.

Reply:
Pushed another change. I implemented `GenerateFalsePositiveExamples` to get test coverage.

Reply:
Beautiful. Ship it. I am frequently asked why there isn't a single converged library of regexes to serve all scenarios but there are some pernicious subtleties in their application that makes it hard to organize patterns to serve all needs. This discussion is a good example.